src/cephadm/cephadm

```diff
@@ -3439,6 +3453,186 @@ class CephContainer:
+
+        unit_file_path = os.path.join(self.ctx.unit_dir, self.unit_name())
+        with open(unit_file_path + '.new', 'w') as f:
+            f.write(self.unit_file())
+            os.rename(unit_file_path + '.new', unit_file_path)
+
+        call_throws(self.ctx, ['systemctl', 'daemon-reload'])
+        call(self.ctx, ['systemctl', 'stop', self.unit_name()],
+             verbosity=CallVerbosity.DEBUG)
+        call(self.ctx, ['systemctl', 'reset-failed', self.unit_name()],
+             verbosity=CallVerbosity.DEBUG)
+        call_throws(self.ctx, ['systemctl', 'enable', '--now', self.unit_name()])
+
+    def unit_name(self) -> str:
+        return '{}.service'.format(get_unit_name(self.fsid, self.daemon_type, self.daemon_id))
+
+    def unit_run(self) -> str:
+        py3 = shutil.which('python3')
+        return ('set -e\n' + f'{py3} {self.binary_path} agent --fsid {self.fsid} --daemon-id {self.daemon_id} &\n')
+
+    def unit_file(self) -> str:
+        return """#generated by cephadm
+[Unit]
+Description=cephadm agent for cluster {fsid}
+
+PartOf=ceph-{fsid}.target
+Before=ceph-{fsid}.target
+
+[Service]
+Type=forking
+ExecStart=/bin/bash {data_dir}/unit.run
+Restart=on-failure
+RestartSec=10s
+
+[Install]
+WantedBy=ceph-{fsid}.target
+""".format(
+            fsid=self.fsid,
+            data_dir=self.daemon_dir
+        )
+
+    def shutdown(self) -> None:
+        self.stop = True
+        if self.mgr_listener.is_alive():
+            self.mgr_listener.shutdown()
+
+    def run(self) -> None:
+        try:
+            with open(self.config_path, 'r') as f:
+                config = json.load(f)
+                self.target_ip = config['target_ip']
+                self.target_port = config['target_port']
+                self.loop_interval = int(config['refresh_period'])
+                self.listener_port = config['listener_port']
+        except Exception as e:
+            self.shutdown()
+            raise Error(f'Failed to get agent target ip and port from config: {e}')
+
+        try:
+            with open(self.keyring_path, 'r') as f:
+                self.keyring = f.read()
+        except Exception as e:
+            self.shutdown()
+            raise Error(f'Failed to get agent keyring: {e}')
+
+        assert self.target_ip and self.target_port
+
+        if not self.mgr_listener.is_alive():
+            self.mgr_listener.start()
+
+        while not self.stop:
+            ack = self.ack
+
+            ls = list_daemons(self.ctx)
+            ls_str = json.dumps(ls)
+
+            facts = HostFacts(self.ctx)
+            facts_str = facts.dump()
+
+            networks = list_networks(self.ctx)
+            networks_str = json.dumps(networks)
+
+            data = {'host': get_hostname(),
+                    'ls': ls_str,
+                    'networks': networks_str,
+                    'facts': facts_str,
+                    'ack': str(ack),
+                    'keyring': self.keyring}
```

Comment:
do you think you should provide the scrape time in the payload sent to the endpoint? If you do that, you could check that against the expected refresh interval to potentially warn about agents falling behind?

Reply:
It's already generating a timestamp on the mgr side when the metadata is received (using it to tell if the agent is "offline" which is currently that it hasn't reported in 2.5x its refresh rate). Do you think we can just stick with timestamps in the mgr or do we need one from the agent itself?